### PR TITLE
Implement source stream concept

### PR DESCRIPTION
**Pull request by jm9e**

### CODE REVIEW

File: builtin/aggregate_test.go [hunk 1]
```diff
@@ -48,11 +51,10 @@ func TestBuiltinAggregate__PassOtherMarkers(t *testing.T) {
...
 	ao.SetParent(do)
 
-	ao.Delegate("iteration").Out().Stream().Map("state").Connect(ao.Delegate("iteration").In().Stream())
-
-	require.NoError(t, do.In().Stream().Map("init").Connect(ao.In().Map("init")))
-	require.NoError(t, do.In().Stream().Map("items").Connect(ao.In().Map("items")))
-	require.NoError(t, ao.Out().Connect(do.Out().Stream()))
+	r.NoError(do.In().Stream().Map("init").Connect(ao.In().Map("init")))
+	r.NoError(do.In().Stream().Map("items").Connect(ao.In().Map("items")))
+	r.NoError(ao.Delegate("iteration").Out().Stream().Map("state").Connect(ao.Delegate("iteration").In().Stream()))
```
Comment (jm9e) [inline builtin/aggregate_test.go:56]: Order of connection is very important now: the source stream "propagates" through operators

File: builtin/aggregate.go [hunk 2]
```diff
@@ -95,4 +95,12 @@ var aggregateOpCfg = &builtinConfig{
...
 		}
 	},
+	oConnFunc: func(dest, src *core.Port) error {
+		o := dest.Operator()
+		if dest == o.In().Map("items") {
+			iOut := o.Delegate("iteration").Out()
+			iOut.SetStreamSource(src.StreamSource())
+		}
+		return nil
+	},
```
Comment (jm9e) [inline builtin/aggregate.go:105]: We make sure that the stream source propagates through the aggregate operator.
See below to understand what it does exactly. This method is being called when an in port of `aggregate` is connected.

File: builtin/loop_test.go [hunk 3]
```diff
@@ -18,7 +18,7 @@ func TestBuiltin_Loop__Simple(t *testing.T) {
 	a := assertions.New(t)
```
Comment (jm9e) [inline builtin/loop_test.go:18]: Basically nothing has changed in this file...

File: builtin/manager.go [hunk 4]
```diff
@@ -63,7 +63,7 @@ func MakeOperator(def core.InstanceDef) (*core.Operator, error) {
 		}
 	}
 
-	o, err := core.NewOperator(def.Name, cfg.oFunc, in, out, dels)
+	o, err := core.NewOperator(def.Name, cfg.oFunc, cfg.oConnFunc, in, out, dels)
```
Comment (jm9e) [inline builtin/manager.go:66]: We have a connection function now, which is a callback for when an `in` port is connected

File: core/port.go [hunk 5]
```diff
@@ -39,6 +39,7 @@ type Port struct {
 	delegate  *Delegate
 	dests     map[*Port]bool
 	src       *Port
+	strSrc    *Port
```
Comment (jm9e) [inline core/port.go:42]: The stream source (a simple port)

File: core/port.go [hunk 6]
```diff
@@ -588,15 +598,23 @@ func setParentStreams(p *Port, parent *Port) {
...
 
+func (p *Port) wire(q *Port) {
+	p.dests[q] = true
+	q.src = p
+	q.strSrc = p.strSrc
+	if q.operator != nil && q.operator.connectFunc != nil {
+		q.operator.connectFunc(q, p)
+	}
+}
+
```
Comment (jm9e) [inline core/port.go:613]: The linear connection code got a bit larger, so we move it out to a new function (moved out from `connect` method)
We also call our new callback method here

File: slang.go [hunk 7]
```diff
@@ -408,11 +408,11 @@ func connectDestinations(o *core.Operator, conns map[*core.Port][]*core.Port) er
...
-			for _, op := range ops {
-				if err := connectDestinations(op, conns); err != nil {
-					return err
-				}
-			}
+		}
+	}
+	for _, op := range ops {
+		if err := connectDestinations(op, conns); err != nil {
+			return err
```
Comment (jm9e) [inline slang.go]: As stated earlier, connection order is very important now! We must not follow connections until **all** `in` ports have been connected. So we move that loop to the end.

File: builtin/aggregate.go [hunk 8]
```diff
@@ -95,4 +95,12 @@ var aggregateOpCfg = &builtinConfig{
 			}
 		}
 	},
+	oConnFunc: func(dest, src *core.Port) error {
+		o := dest.Operator()
+		if dest == o.In().Map("items") {
```
Comment (jm9e) [inline builtin/aggregate.go:100]: This is the `in` port which receives the items from outside. This stream should simply propagate through the aggregator.

File: builtin/aggregate.go [hunk 9]
```diff
@@ -95,4 +95,12 @@ var aggregateOpCfg = &builtinConfig{
...
+			iOut := o.Delegate("iteration").Out()
+			iOut.SetStreamSource(src.StreamSource())
```
Comment (jm9e) [inline builtin/aggregate.go:102]: Hence we set the stream source of the `out` port of our delegate to the same port as the stream source of the `in` port.

**Review by td5r — APPROVED** [review]

**Pull request merged**